clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/PartitionTracker.java

```diff
@@ -19,136 +20,178 @@
   }
 
-  public ReentrantLock getPartitionLock(int partition) {
-    return partitionLocks.computeIfAbsent(partition, key -> new ReentrantLock());
+  /** N.B. Intended for tests */
+  Set<GUID> getTrackedGUIDs() {
+    return Collections.unmodifiableSet(this.segments.keySet());
   }
 
   /**
-   * In some cases, such as when resetting offsets or unsubscribing from a partition,
-   * the {@link ProducerTracker} should forget about the state that it accumulated
-   * for a given partition.
-   *
-   * @param partition to clear state for
+   * @param guid for which to retrieve the lock and segment
+   * @return a non-null {@link ReentrantLockWithRef} (but which can have {@link ReentrantLockWithRef#containsRef()} == false)
    */
-  public void clearPartition(int partition) {
-    ReentrantLock partitionLock = getPartitionLock(partition);
-    partitionLock.lock();
-    try {
-      segments.remove(partition);
-    } finally {
-      partitionLock.unlock();
-    }
+  private ReentrantLockWithRef<Segment> getLockAndSegment(GUID guid) {
+    return this.segments.computeIfAbsent(guid, segmentProvider);
+  }
+
+  /**
+   * @param guid for which to retrieve the lock and segment
+   * @return a {@link ReentrantLockWithRef} or null if it's absent
+   */
+  ReentrantLockWithRef<Segment> getLockAndSegmentIfPresent(GUID guid) {
+    return this.segments.get(guid);
   }
 
-  public void setPartitionState(int partition, ProducerPartitionState state) {
-    setPartitionState(partition, new Segment(partition, state));
+  public void setPartitionState(OffsetRecord offsetRecord, long maxAgeInMs) {
+    long minimumRequiredRecordProducerTimestamp =
+        maxAgeInMs == DISABLED ? DISABLED : offsetRecord.getMaxMessageTimeInMs() - maxAgeInMs;
+    Iterator<Map.Entry<CharSequence, ProducerPartitionState>> iterator =
+        offsetRecord.getProducerPartitionStateMap().entrySet().iterator();
+    Map.Entry<CharSequence, ProducerPartitionState> entry;
+    GUID producerGuid;
+    ProducerPartitionState producerPartitionState;
+    while (iterator.hasNext()) {
+      entry = iterator.next();
+      producerGuid = GuidUtils.getGuidFromCharSequence(entry.getKey());
+      producerPartitionState = entry.getValue();
+      if (producerPartitionState.messageTimestamp >= minimumRequiredRecordProducerTimestamp) {
+        /**
+         * This {@link producerPartitionState} is eligible to be retained, so we'll set the state in the
+         * {@link PartitionTracker}.
+         */
+        setPartitionState(producerGuid, new Segment(partition, producerPartitionState));
+      } else {
+        // The state is eligible to be cleared.
+        segments.remove(producerGuid);
+        iterator.remove();
+      }
+    }
   }
 
-  private void setPartitionState(int partition, Segment segment) {
-    ReentrantLock partitionLock = getPartitionLock(partition);
-    partitionLock.lock();
+  private void setPartitionState(GUID guid, Segment segment) {
```

Comment:
We may want to rename it to be `setProducerState` to reflect the function purpose.

Reply:
Thanks. Good call. I renamed it to `setSegment`.